src/yoroi-wallets/cardano/ShelleyWallet.ts

```diff
@@ -109,227 +116,259 @@ export class ShelleyWallet implements WalletInterface {
+    storage: typeof storageLegacy
+    internalChain: AddressChain
+    externalChain: AddressChain
+
+    isReadOnly: boolean
+  }) => {
+    const rewardAddressHex = await deriveRewardAddressHex(accountPubKeyHex, networkId)
+    const apiUrl = getCardanoNetworkConfigById(networkId).BACKEND.API_ROOT
+    const utxoManager = await makeUtxoManager(id, apiUrl, storage)
+    const transactionCache = await TransactionCache.create(makeStorageWithPrefix(`/wallet/${id}/txs`))
 
-    this.rewardAddressHex = await deriveRewardAddressHex(accountPubKeyHex, networkId)
+    const wallet = new ShelleyWallet({
+      storage,
+      networkId,
+      id,
+      utxoManager,
+      implementationId,
+      hwDeviceInfo,
+      isReadOnly,
+      provider: undefined,
+      accountPubKeyHex,
+      rewardAddressHex,
+      transactionCache,
+      internalChain,
+      externalChain,
+    })
 
+    await wallet.discoverAddresses()
+    wallet.setupSubscriptions()
+    wallet.notify({type: 'initialize'})
+    wallet.isInitialized = true
+
+    if (!isYoroiWallet(wallet)) throw new Error('invalid wallet')
+    return wallet
+  }
+
+  private constructor({
+    storage,
+    networkId,
+    id,
+    utxoManager,
+    implementationId,
+    hwDeviceInfo,
+    isReadOnly,
+    provider,
+    accountPubKeyHex,
+    rewardAddressHex,
+    transactionCache,
+    internalChain,
+    externalChain,
+  }: {
+    storage: typeof storageLegacy
+    networkId: NetworkId
+    id: string
+    utxoManager: UtxoManager
+    implementationId: WalletImplementationId
+    hwDeviceInfo: HWDeviceInfo | null
+    isReadOnly: boolean
+    provider: YoroiProvider | null | undefined
+    accountPubKeyHex: string
+    rewardAddressHex: string
+    transactionCache: TransactionCache
+    internalChain: AddressChain
+    externalChain: AddressChain
+  }) {
+    this.id = id
+    this.storage = storage
+    this.networkId = networkId
+    this.defaultAsset = getDefaultAssetByNetworkId(this.networkId)
+    this.utxoManager = utxoManager
+    this._utxos = utxoManager.initialUtxos
+    this.encryptedStorage = makeWalletEncryptedStorage(id)
+    this.walletImplementationId = implementationId
+    this.isHW = hwDeviceInfo != null
+    this.hwDeviceInfo = hwDeviceInfo
+    this.isReadOnly = isReadOnly
+    this.provider = provider
+    this.transactionCache = transactionCache
+    this.internalChain = internalChain
+    this.externalChain = externalChain
+    this.rewardAddressHex = rewardAddressHex
     this.publicKeyHex = accountPubKeyHex
-
     this.version = DeviceInfo.getVersion()
-
     this.checksum = isByron(implementationId)
       ? legacyWalletChecksum(accountPubKeyHex)
       : walletChecksum(accountPubKeyHex)
-
-    // Create at least one address in each block
-    await this.internalChain.initialize()
-    await this.externalChain.initialize()
-
-    await this.discoverAddresses()
-
     this.setupSubscriptions()
     this.notify({type: 'initialize'})
-
     this.isInitialized = true
```

Comment:
? maybe
```suggestion
    this.initialized = true
```

Reply:
lets keep it for now
most of the yoroi booleans are `isFoo` anyway